### PR TITLE
feat: auto redirection based on user's signed in status

### LINKED ISSUE #132: [FEATURE] <If a user is not signed in on the question submission page, redirect to signup/signin>
### Description

If the user is not signed in at /submit, they should be redirected.

### PR DESCRIPTION
**Pull request by ykdojo**
- If a user's signed in and is on /signin or /signup, they're redirected to / (index)
- If a user's NOT signed in and is on /submit, they're redirected to signup

Closes #132.

### CODE REVIEW

File: src/pages/signin.tsx [hunk 1]
```diff
@@ -3,9 +3,21 @@ import Head from 'next/head';
 import Link from 'next/link';
 import { type NextPage } from 'next';
 import { SignInForm } from '../components/SignIn/Form';
-// import { SignInSocial } from '../components/SignIn/Social';
+import { useEffect } from 'react';
+import Router from 'next/router';
+import { type UserProps } from '../lib/types';
+
+const SignIn: NextPage<UserProps> = (props: UserProps) => {
```
Comment (subhoghoshX) [inline src/pages/signin.tsx:10]: I don't think we need to specify `UserProps` twice here.

**Comment by subhoghoshX** [comment]
Thank you. One thing I don't like is that when a user is logged in and they visit `/signin` or `/signup` there's a flush of the page before it redirects. I think it'd be a better UX if we could redirect the user before that happens.

**Comment by subhoghoshX** [comment]
Thank you. One thing, when a user is logged in and they visit `/signin` or `/signup` there's a flush of the page before it redirects. I think it'd be a better UX if we could redirect the user before that happens.

File: src/pages/_app.tsx [hunk 2]
```diff
@@ -1,8 +1,44 @@
...
+		async function getInitialSession() {
+			const supabaseSession = await supabase.auth.session();
+
+			// only update the react state if the component is still mounted
+			if (mounted) {
+				if (supabaseSession) {
+					setSession(supabaseSession);
+				}
+			}
+		}
```
Comment (subhoghoshX) [inline src/pages/_app.tsx:26]: How about something like:
```
if (mounted && supabaseSession) {
    setSession(supabaseSession);
}
```

File: src/pages/signin.tsx (hunk 1 above)
Comment (subhoghoshX) [reply src/pages/signin.tsx:10]: And we could also destructure the user:
```js
const SignIn: NextPage<UserProps> = ({ user }: UserProps) => {
```
That way won't need the below line, i.e.
```js
const user = props.user;
```

**Review by emperorsegxy — APPROVED** [review]
Everything looks great.

File: src/pages/signin.tsx (hunk 1 above)
Comment (ykdojo) [reply src/pages/signin.tsx:10]: done

**Comment by ykdojo** [comment]
> Thank you. One thing, when a user is logged in and they visit /signin or /signup there's a flush of the page before it redirects. I think it'd be a better UX if we could redirect the user before that happens.

The thing is Supabase seems to take a moment to load the user. Even if we use localStorage like I did in /submit, React/Next.js doesn't seem to load it right away.

Plus, a link to the signin/singup page shouldn't show up anywhere on the site if the user's already signed in, so it's more of a safety measure.

File: src/pages/_app.tsx (hunk 2 above)
Comment (ykdojo) [reply src/pages/_app.tsx:26]: done, good idea

**Comment by subhoghoshX** [comment]
> > Thank you. One thing, when a user is logged in and they visit /signin or /signup there's a flush of the page before it redirects. I think it'd be a better UX if we could redirect the user before that happens.
> 
> The thing is Supabase seems to take a moment to load the user. Even if we use localStorage like I did in /submit, React/Next.js doesn't seem to load it right away.
> 
> Plus, a link to the signin/singup page shouldn't show up anywhere on the site if the user's already signed in, so it's more of a safety measure.

Yes. Never mind. I was just nitpicking there. It's pretty common in client-side routing. And I don't think we can do much there without delaying the rending of the component. Which I think would also hurt the performance.

Thank you. It looks good now. We can merge it.

**Comment by subhoghoshX** [comment]
Oops! Looks like there's still the extra `UserProps` in submit and signup page.

Let me try something new.

**Comment by ykdojo** [comment]
haha no worries, and it's done

**Pull request merged**